### PR TITLE
mgr/dashboard: add rbd status endpoint and error page

### PR DESCRIPTION
**Pull request by melissa-kun-li**
Show "No RBD pools available" error page when accessing block/rbd if there are no rbd pools. Also, add a "button_name" and "button_route" property to `ModuleStatusGuardService` config to customize the button on the error page (the default button is "Go To Dashboard").
This PR will also modify `ModuleStatusGuardService` to execute API calls to `/ui-api/<uiApiPath>/status` which uses the `UIRouter`.

rbd-status-endpoint

Screenshot from 2022-05-31 19-43-24

Fixes: https://tracker.ceph.com/issues/42109
Signed-off-by: Melissa Li <melissali@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by epuertat — COMMENTED** [review]
Great first Dashboard PR @melissa-kun-li! You even dared with observables 😜  !

I'm curious, what decided you to finally go with this implementation? I initially mentioned this one, but after our conversation with you and Avan, you brought up the [empty state pattern](https://www.patternfly.org/v4/components/empty-state/) and loved that idea.

I personally find the "empty state" better suited for this case (in fact we're applying that to other components: orchestrator, NFS, RGW, ...). Keep in mind that a user cannot do anything in that page until the first RBD pool is created, but we are displaying so much information (a table, buttons, filters, etc.). So why not directly guide the user towards that? The empty state is an excellent option for wizards and guiding users to day-1 (post-install) actions.

image

This pattern that you implemented here (disabling the button) would work great when the user is already familar with the RBD page (day-2, maintenance ops) and we don't want them to make mistakes (or clicking buttons that would lead them to unproductive actions).

A more minimalistic alternative to this would have been to hide those buttons, but that's sometimes considered not a good practice, since by presenting the disabled buttons the user knows that they need to do something else to enable those.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.ts [hunk 1]
```diff
@@ -69,6 +71,8 @@ export class RbdListComponent extends ListWithDetails implements OnInit {
   tableStatus = new TableStatusViewCache();
   selection = new CdTableSelection();
   icons = Icons;
+  rbdPools: Array<RbdPool>;
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.ts:74]: You may change it to boolean value as just checking if any pools exists or not is enough!
```suggestion
  rbdPoolsExist: boolean;
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.ts [hunk 2]
```diff
@@ -307,6 +312,10 @@ export class RbdListComponent extends ListWithDetails implements OnInit {
       itemFilter,
       this.builders
     );
+
+    this.rbdPoolSubscription = this.rbdService.list().subscribe((resp: Array<RbdPool>) => {
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.ts:316]: The subscription to this service is already done [here](https://github.com/ceph/ceph/blob/fd3a9eb92f807e8f4571c2436c431595e41e2dd4/src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.ts#L306)  and it's getting processed in the next line, 
`      (resp) => this.prepareResponse(resp),`
So you can probably add a extra line inside function `this.rbdPools = resp` or I think a boolean  is enough too.
`rbdPoolsExist = !(resp === undefined || resp.length == 0)`

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.ts [hunk 3]
```diff
@@ -477,4 +486,17 @@ export class RbdListComponent extends ListWithDetails implements OnInit {
     }
     return false;
   }
+
+  getCreateDisableDesc(): string | boolean {
+    if (this.rbdPools !== undefined && this.rbdPools.length === 0) {
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.ts:491]:
```suggestion
    if (!this.rbdPoolsExist) {
```

**Comment by melissa-kun-li** [comment]
> Great first Dashboard PR @melissa-kun-li! You even dared with observables stuck_out_tongue_winking_eye !
> 
> I'm curious, what decided you to finally go with this implementation? I initially mentioned this one, but after our conversation with you and Avan, you brought up the [empty state pattern](https://www.patternfly.org/v4/components/empty-state/) and loved that idea.
>
Haha thanks @epuertat ! So originally I was planning on going with this implementation, but after I talked with @avanthakkar it seemed like it wouldn't be possible to implement it from a code perspective, so he thought of the empty state pattern as an alternative, which I thought was a good idea. But later I talked with Avan again (after talking with you about the empty state pattern), and Avan said that it actually was indeed possible to go with this "disable Create button" approach. 

What made me choose this approach was when I was thinking about what would happen if a user: 1) had an rbd pool 2) created an rbd image using that rbd pool 3) deleted their rbd pool for whatever reason, maybe by accident 4) tried to access the Images table view.

I wasn't sure what would happen in that scenario - I assumed that it if we went with the empty state pattern, it would show that "No rbd pools available" page that kinda blocks the table view. But if we went with the approach of disabling the "Create" button, I assumed it would still allow the user to access the "Edit", "Delete", etc. options that are in the table view, so it maybe could prevent the image from being stuck in error state of not having an rbd pool.

That was my thought process behind it, what do you think? Also, if empty state pattern would be better suited, I can definitely pivot to that. Learning how to do this was interesting in itself :D

**Review by epuertat — APPROVED** [review]
LGTM. I only checked the code (not tested it actually) :D

**Comment by epuertat** [comment]
@melissa-kun-li there are 4 make check issues:

- [projectroot.src.pybind.mgr.dashboard.run_tox_mgr_dashboard_openapi](https://jenkins.ceph.com/job/ceph-pull-requests/96477/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_openapi/)
- [projectroot.src.pybind.mgr.dashboard.run_tox_mgr_dashboard_lint](https://jenkins.ceph.com/job/ceph-pull-requests/96477/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_lint/)
- [projectroot.src.test.mgr.mgr_dashboard_frontend_unittests](https://jenkins.ceph.com/job/ceph-pull-requests/96477/testReport/junit/projectroot.src.test/mgr/mgr_dashboard_frontend_unittests/)
- [projectroot.src.pybind.mgr.run_tox_mgr](https://jenkins.ceph.com/job/ceph-pull-requests/96477/testReport/junit/projectroot.src.pybind/mgr/run_tox_mgr/)

File: src/pybind/mgr/dashboard/controllers/rbd.py [hunk 4]
```diff
@@ -81,6 +81,16 @@ class MIRROR_IMAGE_MODE(Enum):
...
 
+    @EndpointDoc("Display RBD Image feature status")
+    @Endpoint()
+    @ReadPermission
+    def status(self):
+        status = {'available': True, 'message': None}
+        if not CephService.get_pool_list('rbd'):
+            status['available'] = False
+            status['message'] = 'No RBD pools in the cluster. Please create a pool with the "rbd" application label.'
+        return status
```
[the remark this answers is not in the document]
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/rbd.py:92]: +1 for this!

**Review by avanthakkar — APPROVED** [review]
Tested locally and works like a charm! Great work @melissa-kun-li 
If you can add uni-test for status endpoint and if possible e2e test to verify workflow would be nice.

File: src/pybind/mgr/dashboard/controllers/rbd.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (melissa-kun-li) [reply src/pybind/mgr/dashboard/controllers/rbd.py:92]: @epuertat can you explain a bit more about the UI endpoints - specifically how to use `UIAPIRouter`? I'm a little confused about what change to make - would it involve using @UIRouter?: https://github.com/ceph/ceph/blob/6fee777d603aebce492c57b41f3b5760d50ddb07/src/pybind/mgr/dashboard/controllers/_ui_router.py#L4

[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/rbd.py]: @melissa-kun-li Dashboard currently exposes 2 kind of REST endpoints: API and UI ones.

* API, under `/api` path, are versioned (there's a version field in the HTTP `Content-Type` header that we honor, so the REST API consumer has to state which version of the API means to use, and if there's an API mismatch, the Dashboard REST API returns [a 415 error](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/415). The goal of this is to maintain a stable API throughout Ceph releases that can be consumed by users for automation/scripting purposes beyond regular UI usage.
* UI endpoints, under `/ui-api`, are not versioned, and might be subject to change in syntax and semantics with no prior warning, and not meant in any to be consumed by users. It's only for the UI layer to obtain ancillary/helper data from the back-end.

Now, about the `Router` - `Controller` architecture: in application servers (webservers that don't merely serve static assets), `Controller` classes handle user-fed data or return data to users based on user-provided directions (usually from/to a back-end data store), while the `Router` class connects that logic to a user-reachable address (that is usually a URL). The Router can also define the conditions under which a user can consume that controller (e.g.: RBAC, permissions, etc.).

**Comment by epuertat** [comment]
[OpenAPI error:](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/10762/console)
```diff
openapi-check run-test: commands[1] | diff openapi.yaml /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/.tox/openapi-check/tmp/openapi.yaml
148,181d147
<   /api/block/image/status:
<     get:
<       parameters: []
<       responses:
<         '200':
<           content:
<             application/vnd.ceph.api.v1.0+json:
<               schema:
<                 properties:
<                   available:
<                     description: RBD Block Image status
<                     type: boolean
<                   message:
<                     description: Error message
<                     type: string
<                 required:
<                 - available
<                 - message
<                 type: object
<           description: OK
<         '400':
<           description: Operation exception. Please check the response body for details.
<         '401':
<           description: Unauthenticated access. Please login first.
<         '403':
<           description: Unauthorized access. Please check your permissions.
<         '500':
<           description: Unexpected error. Please check the response body for the stack
<             trace.
<       security:
<       - jwt: []
<       summary: Status of RBD Block Image feature
<       tags:
<       - RBD-Block-Image
328a295,317
>       tags:
>       - Rbd
>   /api/block/image/status:
>     get:
>       parameters: []
>       responses:
>         '200':
>           content:
>             application/vnd.ceph.api.v1.0+json:
>               type: object
>           description: OK
>         '400':
>           description: Operation exception. Please check the response body for details.
>         '401':
>           description: Unauthenticated access. Please login first.
>         '403':
>           description: Unauthorized access. Please check your permissions.
>         '500':
>           description: Unexpected error. Please check the response body for the stack
>             trace.
>       security:
>       - jwt: []
>       summary: Display RBD Image feature status
```

**Comment by epuertat** [comment]
e2e (this is a weird unrelated error that sometimes happens due to some conflict with conflicting versions of the ceph_argparse file, probably because of some undeleted orphan file from a previous PR testing):
```
2022-06-03T05:07:36.379+0000 7f55f4a4f100 -1 mgr[py] Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/src/pybind/mgr/telemetry/__init__.py", line 7, in <module>
    from .module import Module
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/src/pybind/mgr/telemetry/module.py", line 123, in <module>
    class Module(MgrModule):
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/src/pybind/mgr/telemetry/module.py", line 1478, in Module
    def on(self, license: Optional[str] = None) -> Tuple[int, str, str]:
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/src/pybind/mgr/mgr_module.py", line 416, in __call__
    self.store_func_metadata(func)
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/src/pybind/mgr/mgr_module.py", line 413, in store_func_metadata
    self._load_func_metadata(f)
  File "/home/jenkins-build/build/workspace/ceph-dashboard-pull-requests/src/pybind/mgr/mgr_module.py", line 397, in _load_func_metadata
    args.append(CephArgtype.to_argdesc(arg_spec[arg],
TypeError: to_argdesc() takes from 2 to 3 positional arguments but 4 were given
```

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
[run_tox_mgr_dashboard_lint](https://jenkins.ceph.com/job/ceph-pull-requests/96707/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_lint/
):
```
lint run-test: commands[2] | isort . --check
ERROR: /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/controllers/rbd.py Imports are incorrectly sorted and/or formatted.
```
This can be fixed with a `tox -e fix` from the dashboard dir.

**Comment by epuertat** [comment]
[run_tox_mgr](https://jenkins.ceph.com/job/ceph-pull-requests/96707/testReport/junit/projectroot.src.pybind/mgr/run_tox_mgr/):
```
dashboard/plugins/feature_toggles.py:12: note: In module imported here,
dashboard/module.py:55: note: ... from here,
dashboard/__init__.py:52: note: ... from here:
dashboard/controllers/rbd.py: note: In member "status" of class "Rbd":
dashboard/controllers/rbd.py:91: error: Incompatible types in assignment (expression has type "str", target has type "Optional[bool]")
Found 1 error in 1 file (checked 31 source files)
```

File: src/pybind/mgr/dashboard/controllers/rbd.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (melissa-kun-li) [reply src/pybind/mgr/dashboard/controllers/rbd.py]: Thanks for the explanation Ernesto! :smiley: I'll try to make a UI endpoint

**Comment by melissa-kun-li** [comment]
jenkins test dashboard cephadm

**Comment by melissa-kun-li** [comment]
@epuertat @avanthakkar  To use UI endpoint, I had to change a few other things as well, would this be a good approach?

To use UI endpoint, I modified `ModuleStatusGuardConfig` to do API calls to `/ui-api/<uiApiPath>/status` and changed a few other classes to `UIRouter` instead of `APIRouter`, since they also use `ModuleStatusGuardConfig` to show an error page. The other error pages are: Orchestrator is not available error page: https://github.com/ceph/ceph/blob/6fee777d603aebce492c57b41f3b5760d50ddb07/src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts#L126-L135, object gateway is not configured (rgw): https://github.com/ceph/ceph/blob/6fee777d603aebce492c57b41f3b5760d50ddb07/src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts#L299-L306, and nfs-ganesha is not configured: https://github.com/ceph/ceph/blob/6fee777d603aebce492c57b41f3b5760d50ddb07/src/pybind/mgr/dashboard/frontend/src/app/app-routing.module.ts#L336-L342

Since I changed those to `/ui-api`, I also removed them from the openapi.yaml

**Comment by epuertat** [comment]
@melissa-kun-li when I was reviewing https://github.com/ceph/ceph/pull/46527 from @nizamial09, I remembered that we had some cross component code to deal with status, but I didn't remember exactly what that was about... I don't think it's a bad idea at all (at least it's consistent), and if we want to leverage the empty state pattern, it's a good way to go

**Comment by melissa-kun-li** [comment]
jenkins test make check

**Comment by melissa-kun-li** [comment]
jenkins test make check

**Review by sunilangadi2 — APPROVED** [review]
Thanks, @melissa-kun-li This is a Nice Improvement!

**Review by nizamial09 — APPROVED** [review]
LGTM!

**Pull request merged**